queue/claimresolver.js

```diff
@@ -0,0 +1,243 @@
+  }
+
+  /** Handle advisory message about claim expiration */
+  async handleMessage({taskId, runId, takenUntil, remove}) {
+    // Query for entity for which we have exact rowKey too, limit to 1, and
+    // require that takenUntil matches. This is essentially a conditional load
+    // operation. Note, that this possible because we set takenUntil both in
+    // the task.runs[runId].takenUntil property and in the task.takenUntil
+    // property whenever a task is claimed, reclaimed. And task.takenUntil is
+    // cleared whenever a run is resolved, so this conditional load should
+    // significantly reduce the amount of task entities that we load.
+    var {entries: [task]} = await this.Task.query({
+      taskId:     taskId,     // Matches an exact entity
+      takenUntil: takenUntil  // Load conditionally
+    }, {
+      matchRow:   'exact',    // Validate that we match row key exactly
+      limit:      1           // Load at most one entity, no need to search
+    });
+
+    // If the task doesn't exist, we're done
+    if (!task) {
+      // don't log this as it's fairly common...
+      return remove();
+    }
+
+    // Check if this is the takenUntil we're supposed to be resolving for, if
+    // this check fails, then the conditional load must have failed so we should
+    // alert operator!
+    if (task.takenUntil.getTime() !== takenUntil.getTime()) {
+      debug("[alert-operator] Task takenUntil doesn't match takenUntil from " +
+            "message, taskId: %s, task.takenUntil: %s, message.takenUntil: %s ",
+            taskId, task.takenUntil.toJSON(), takenUntil.toJSON());
+      return remove();
+    }
+
+    // Ensure that all runs are resolved
+    await task.modify((task) => {
+      var run = task.runs[runId];
+      if (!run) {
+        // The run might not have been created, if the claimTask operation
+        // failed
+        debug("[not-a-bug] runId: %s does exists on taskId: %s, but " +
+              "deadline message has arrived", runId, taskId);
+        return;
+      }
+
+      // If the run isn't running, or takenUntil has been updated, then we're
+      // done. Notice that unlike condition above, we're racing in the
+      // task.modify modifier so reloads can happen if there is concurrency!
+      // Hence, takenUntil being update is both valid and plausible.
+      if (run.state !== 'running' ||
+          new Date(run.takenUntil).getTime() !== takenUntil.getTime()) {
+        return;
+      }
+
+      // Update run
+      run.state           = 'exception';
+      run.reasonResolved  = 'claim-expired';
+      run.resolved        = new Date().toJSON();
+
+      // Do **NOT** clear takenUntil on task, as this will prevent the message
+      // from running again. In case something below fails.
+
+      // If the run isn't the last run, then something is very wrong
+      if (task.runs.length - 1 !== runId) {
+        debug("[alert-operator] running runId: %s, resolved exception, " +
```

Comment:
what else can we do... Crash isn't better...

It's an invariant violation. I seriously doubt it's going to happen... If it's happens it's probably a freak task that we want to investigate. If happens all the time we're well not good no matter what we do :)